// probot/lib/context.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var deepmerge_1 = __importDefault(require("deepmerge"));
var js_yaml_1 = __importDefault(require("js-yaml"));
var path_1 = __importDefault(require("path"));
var CONFIG_PATH = '.github';
var BASE_KEY = '_extends';
var BASE_REGEX = new RegExp('^' +
    '(?:([a-z\\d](?:[a-z\\d]|-(?=[a-z\\d])){0,38})/)?' + // org
    '([-_.\\w\\d]+)' + // project
    '(?::([-_./\\w\\d]+\\.ya?ml))?' + // filename
    '$', 'i');
var DEFAULT_BASE = '.github';
/**
 * The context of the event that was triggered, including the payload and
 * helpers for extracting information can be passed to GitHub API calls.
 *
 *  ```js
 *  module.exports = app => {
 *    app.on('push', context => {
 *      context.log('Code was pushed to the repo, what should we do with it?');
 *    });
 *  };
 *  ```
 *
 * @property {github} github - A GitHub API client
 * @property {payload} payload - The webhook event payload
 * @property {logger} log - A logger
 */
var Context = /** @class */ (function () {
    function Context(event, github, log) {
        this.name = event.name;
        this.id = event.id;
        this.payload = event.payload;
        this.protocol = event.protocol;
        this.host = event.host;
        this.url = event.url;
        this.github = github;
        this.log = log;
    }
    Object.defineProperty(Context.prototype, "event", {
        // Maintain backward compatibility
        get: function () {
            return this.name;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Return the `owner` and `repo` params for making API requests against a
     * repository.
     *
     * ```js
     * const params = context.repo({path: '.github/config.yml'})
     * // Returns: {owner: 'username', repo: 'reponame', path: '.github/config.yml'}
     * ```
     *
     * @param object - Params to be merged with the repo params.
     *
     */
    Context.prototype.repo = function (object) {
        var repo = this.payload.repository;
        if (!repo) {
            throw new Error('context.repo() is not supported for this webhook event.');
        }
        return Object.assign({
            owner: repo.owner.login || repo.owner.name,
            repo: repo.name
        }, object);
    };
    /**
     * Return the `owner`, `repo`, and `number` params for making API requests
     * against an issue or pull request. The object passed in will be merged with
     * the repo params.
     *
     * ```js
     * const params = context.issue({body: 'Hello World!'})
     * // Returns: {owner: 'username', repo: 'reponame', number: 123, body: 'Hello World!'}
     * ```
     *
     * @param object - Params to be merged with the issue params.
     */
    Context.prototype.issue = function (object) {
        var payload = this.payload;
        return Object.assign({
            number: (payload.issue || payload.pull_request || payload).number
        }, this.repo(object));
    };
    Object.defineProperty(Context.prototype, "isBot", {
        /**
         * Returns a boolean if the actor on the event was a bot.
         * @type {boolean}
         */
        get: function () {
            return this.payload.sender.type === 'Bot';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Reads the app configuration from the given YAML file in the `.github`
     * directory of the repository.
     *
     * For example, given a file named `.github/config.yml`:
     *
     * ```yml
     * close: true
     * comment: Check the specs on the rotary girder.
     * ```
     *
     * Your app can read that file from the target repository:
     *
     * ```js
     * // Load config from .github/config.yml in the repository
     * const config = await context.config('config.yml')
     *
     * if (config.close) {
     *   context.github.issues.comment(context.issue({body: config.comment}))
     *   context.github.issues.edit(context.issue({state: 'closed'}))
     * }
     * ```
     *
     * You can also use a `defaultConfig` object:
     *
     * ```js
     * // Load config from .github/config.yml in the repository and combine with default config
     * const config = await context.config('config.yml', {comment: 'Make sure to check all the specs.'})
     *
     * if (config.close) {
     *   context.github.issues.comment(context.issue({body: config.comment}));
     *   context.github.issues.edit(context.issue({state: 'closed'}))
     * }
     * ```
     *
     * Config files can also specify a base that they extend. `deepMergeOptions` can be used
     * to configure how the target config, extended base, and default configs are merged.
     *
     * @param fileName - Name of the YAML file in the `.github` directory
     * @param defaultConfig - An object of default config options
     * @param deepMergeOptions - Controls merging configs (from the [deepmerge](https://github.com/TehShrike/deepmerge) module)
     * @return Configuration object read from the file
     */
    Context.prototype.config = function (fileName, defaultConfig, deepMergeOptions) {
        return __awaiter(this, void 0, void 0, function () {
            var params, config, baseRepo, baseConfig, baseParams;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        params = this.repo({ path: path_1.default.posix.join(CONFIG_PATH, fileName) });
                        return [4 /*yield*/, this.loadYaml(params)];
                    case 1:
                        config = _a.sent();
                        if (config == null) {
                            baseRepo = DEFAULT_BASE;
                        }
                        else if (config != null && BASE_KEY in config) {
                            baseRepo = config[BASE_KEY];
                            delete config[BASE_KEY];
                        }
                        if (!baseRepo) return [3 /*break*/, 3];
                        if (typeof baseRepo !== 'string') {
                            throw new Error("Invalid repository name in key \"" + BASE_KEY + "\"");
                        }
                        baseParams = this.getBaseParams(params, baseRepo);
                        return [4 /*yield*/, this.loadYaml(baseParams)];
                    case 2:
                        baseConfig = _a.sent();
                        _a.label = 3;
                    case 3:
                        if (config == null && baseConfig == null && !defaultConfig) {
                            return [2 /*return*/, null];
                        }
                        return [2 /*return*/, deepmerge_1.default.all(
                            // filter out null configs
                            [defaultConfig, baseConfig, config].filter(function (conf) { return conf; }), deepMergeOptions)];
                }
            });
        });
    };
    /**
     * Loads a file from GitHub
     *
     * @param params Params to fetch the file with
     * @return The parsed YAML file
     */
    Context.prototype.loadYaml = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var response, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.github.repos.getContents(params)
                            // Ignore in case path is a folder
                            // - https://developer.github.com/v3/repos/contents/#response-if-content-is-a-directory
                        ];
                    case 1:
                        response = _a.sent();
                        // Ignore in case path is a folder
                        // - https://developer.github.com/v3/repos/contents/#response-if-content-is-a-directory
                        if (Array.isArray(response.data)) {
                            return [2 /*return*/, null];
                        }
                        // we don't handle symlinks or submodule
                        // - https://developer.github.com/v3/repos/contents/#response-if-content-is-a-symlink
                        // - https://developer.github.com/v3/repos/contents/#response-if-content-is-a-submodule
                        if (typeof response.data.content !== 'string') {
                            return [2 /*return*/];
                        }
                        return [2 /*return*/, js_yaml_1.default.safeLoad(Buffer.from(response.data.content, 'base64').toString()) || {}];
                    case 2:
                        e_1 = _a.sent();
                        if (e_1.status === 404) {
                            return [2 /*return*/, null];
                        }
                        throw e_1;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Computes parameters for the repository specified in base
     *
     * Base can either be the name of a repository in the same organization or
     * a full slug "organization/repo".
     *
     * @param params An object containing owner, repo and path
     * @param base A string specifying the base repository
     * @return The params of the base configuration
     */
    Context.prototype.getBaseParams = function (params, base) {
        var match = base.match(BASE_REGEX);
        if (match === null) {
            throw new Error("Invalid repository name in key \"" + BASE_KEY + "\": " + base);
        }
        return {
            owner: match[1] || params.owner,
            path: match[3] || params.path,
            repo: match[2]
        };
    };
    return Context;
}());
exports.Context = Context;
